contents/docs/architecture/ingestion-pipeline.mdx

```diff
@@ -0,0 +1,135 @@
+6. [Apps - `onEvent`](#6-apps---onevent)
+
+### 1. Event buffer
+
+The event buffer sits right at the beginning of the ingestion pipeline, and gives us the ability to selectively delay the processing of certain events.
+Events are initially read from the `events_plugin_ingestion` Kafka topic, which is written to by the Capture API.
+
+#### Why are events buffered?
+
+Since events coming in to PostHog can arrive in any order, we sometimes decide to delay the processing of specific events to fix edge cases that can potentially arrive.
+It will become more obvious why this is necessary as we dive deeper into the event pipeline.
+
+#### Determining if an event should be buffered
+
+After an event is conusmed from Kafka, the plugin server will check a number of things in order to determine whether or not to buffer an event.
+If any of these conditions are true the event will not be buffered and will immediately move on to the [next step](#2-apps---processevent).
+
+-   If the `distinct_id` on the event is already associated with an existing Person
+-   If the event is _anonymous_ (In this case this means the `distinct_id` matches the `device_id` generated by the library that sent the event)
+-   If the event is an `$identify` call
+-   If the event is coming from one of our mobile libraries, as it is not easy to determine if an event is _anonymous_ and it is not worth it to buffer all events on mobile
+
+If an event coming in satisfies none of these checks, then it will be added to the buffer and sent back through the ingestion pipeline in _at most_ 60s.
+
+For more detail on this step, check out [this file](https://github.com/PostHog/posthog/blob/master/plugin-server/src/worker/ingestion/event-pipeline/1-emitToBufferStep.ts) from the plugin-server codebase.
+
+### 2. Apps - `processEvent`
+
+After the event buffer, we start the first of a few steps that augment or transform our raw event before it gets written into Clickhouse.
+This first step runs any workloads that come from [Apps](/apps) that you have installed and who have exported a `processEvent` function.
+This is the only chance for apps to transform or exclude an event before it is writtend into Clickhouse.
+
+For more detail on this step, check out [this file](https://github.com/PostHog/posthog/blob/master/plugin-server/src/worker/ingestion/event-pipeline/2-pluginsProcessEventStep.ts) from the plugin-server codebase.
+
+### 3. Person processing
+
+The next step in the ingestion pipeline is processing the Person who sent the event, which is determined by the `distinct_id` field.
+A number of different actions can happend here depending both on if we've seen this `distinct_id` before, as well as which type of event is being sent.
+
+To determine whether or not a Person and corresponding `person_id` have already been created and associated with a `distinct_id`, we make a call to PostgreSQL to retrieve the latest mappings.
+
+#### `$identify` or `$alias` event
```

Comment:
It's useful to think of it as we send two distinct_ids here:
https://github.com/PostHog/posthog/blob/aedba0258660592d26fe0f40a62b090472c992de/plugin-server/src/worker/ingestion/person-state.ts#L272
``` 
identify(distinct_id, event_properties[`anon_distinct_id`])
alias(distinct_id1, distinct_id2)
```
then there's 3 cases: person doesn't exist for either, exists for 1 of them, or exists for both of them (in the table below)

Reply:
Got it, this makes a lot of sense now. I had forgotten about the `$anon_distinct_id` field which made things a bit confusing, so thinking about it in the context of those two fields together makes it much more clear